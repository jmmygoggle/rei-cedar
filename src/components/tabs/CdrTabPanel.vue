<template>
  <div
    :aria-hidden="!active"
    :aria-labelledby="ariaLabelledby"
    :class="componentClass"
    :hidden="hidden"
    :id="id"
    tabindex="0"
    role="tabpanel"
    @keydown_up_prevent="handleUpArrowNav"
    @animationend="animationEnd"
    :key="name"
  >
    <slot />
  </div>
</template>
<script lang="ts">
import { defineComponent, computed, ref } from 'vue';


// TODO: PUT MODIFIER HERE!
import clsx from 'clsx';
import { buildClass } from '../../utils/buildClass';
import propValidator from '../../utils/propValidator';
import style from './styles/CdrTabPanel.scss';

export default {
  name: 'CdrTabPanel',
  props: {
    /**
     * Required string value that shows up on tab header
     */
    name: {
      type: String,
      required: true,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    ariaLabelledBy: {
      type: String,
      required: true,
    },
    id: {
      type: String,
      required: true,
    },
    modifier: {
      type: String,
      default: '',
      validator: (value) => propValidator(value, ['centered', 'compact', 'full-width', 'no-border']),
    },
  },

  setup(props, ctx) {
    const active = ref(false);
    const hidden = ref(true);
    const offsetX = ref(0);
    const animationDirection = ref(null);
    
    return {


// <!-- componentClass: clsx(this.style[this.baseClass], this.modifierClass, this.animationClass) -->


      // computed: {
      //   baseClass() {
      //     return 'cdr-tab-panel';
      //   },
      //   animationClass() {
      //     return this.animationDirection ? style[`cdr-tab-panel-${this.animationDirection}`] : null;
      //   },
      // },
      // methods: {
      //   setActive(state) {
      //     if (state) this.hidden = false;
      //     this.active = state;
      //     this.$emit('tab-change', state, this.id);
      //   },
      //   setAnimationDirection(direction) {
      //     this.animationDirection = direction;
      //   },
      //   setOffsetX(x) {
      //     this.offsetX = x;
      //   },
      //   handleUpArrowNav() {
      //     this.$parent.setFocusToActiveTabHeader();
      //   },
      //   animationEnd(event) {
      //     if (event.animationName.split('-')[0] === 'exit') {
      //       this.hidden = true;
      //       this.animationDirection = null;
      //     }
      //   },
      // },







    };
  },
};

</script>
