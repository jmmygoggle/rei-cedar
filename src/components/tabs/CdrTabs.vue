<template>
  <div
    class={clsx(this.style[this.baseClass], this.modifierClass, this.sizeClass)}
    ref="cdrTabsContainer"
    style={{ height: this.height }}
  >
    <div
      class={this.style['cdr-tabs__gradient-container']}
      vOn:keyup_right={this.rightArrowNav}
      vOn:keyup_left={this.leftArrowNav}
      vOn:keydown_down_prevent={this.handleDownArrowNav}
    >
      <div class={clsx(
        this.style['cdr-tabs__gradient'],
        this.style['cdr-tabs__gradient--left'],
        this.overflowLeft ? this.style['cdr-tabs__gradient--active'] : '',
      )}
        style={this.gradientLeftStyle}
      ></div>
      <nav
        class={this.style['cdr-tabs__header-container']}
      >
        <div
          class={this.style['cdr-tabs__header']}
          role="tablist"
          ref="cdrTabsHeader"
        >
          {this.tabs.map((tab) => this.getTabEl(tab))}

          <!-- return tab.disabled ? (
            <button
              class={clsx(
                this.style['cdr-tabs__header-item'],
                this.style['cdr-tabs__header-item--disabled'],
              )}
              disabled
            >
              {tab.name}
            </button>
          ) : (
            <button
              role="tab"
              aria-selected={tab.active}
              aria-controls={tab.id}
              id={tab.ariaLabelledby}
              key={tab.id}
              class={clsx(
                tab.active ? this.style['cdr-tabs__header-item-active'] : '',
                this.style['cdr-tabs__header-item'],
              )}
              tabIndex={tab.active ? 0 : -1}
              vOn:click_prevent={(e) => this.handleClick(tab, e)}
              js-name={ tab.name }
            >
              { tab.name }
            </button> -->



        </div>
      </nav>
      <div class={clsx(
        this.style['cdr-tabs__gradient'],
        this.style['cdr-tabs__gradient--right'],
        this.overflowRight ? this.style['cdr-tabs__gradient--active'] : '',
      )}
        style={this.gradientRightStyle}
      ></div>
      <div
        class={this.style['cdr-tabs__underline']}
        style={this.underlineStyle}
      />
    </div>
    <div
      class={this.style['cdr-tabs__content-container']}
      ref="slotWrapper"
    >
      {this.$slots.default}
    </div>
  </div>
</template>
<script lang="ts">
import { defineComponent, computed } from 'vue';
import { buildClass } from '../../utils/buildClass'
import propValidator from '../../utils/propValidator';

const modifierClass = computed(() => buildClass('cdr-radio', props.modifier, style));
const sizeClass = computed(() => props.size && buildClass('cdr-radio', props.size, style));



import debounce from 'lodash-es/debounce';
import clsx from 'clsx';
import {
  CdrColorBackgroundPrimary, CdrSpaceOneX, CdrSpaceHalfX,
} from '@rei/cdr-tokens/dist/js/cdr-tokens.esm';
// import modifier from '../../mixins/modifier';
// import size from '../../mixins/size';
import style from './styles/CdrTabs.scss';

export default {
  name: 'CdrTabs',
  mixins: [modifier, size],
  props: {
    height: {
      type: String,
      default: '240px',
    },
    activeTab: {
      type: Number,
      default: 0,
    },
    backgroundColor: {
      type: String,
      default: CdrColorBackgroundPrimary,
    },
  },
  mounted() {
    // TODO: DO something like accordion group here?
    // hrmmmm
    this.tabs = (this.$slots.default || [])
      .map((vnode) => vnode.componentInstance)
      .filter((tab) => tab); // get vue component children in the slot

    this.activeTabIndex = this.getNextTab(this.activeTab);

    if (this.tabs[this.activeTabIndex] && this.tabs[this.activeTabIndex].setActive) {
      this.tabs[this.activeTabIndex].setActive(true);
    }

    this.$nextTick(() => {
      this.headerWidth = this.getHeaderWidth();
      this.calculateOverflow();
      setTimeout(() => {
        this.updateUnderline();
      }, 500);
    });
    // Check for header overflow on window resize for gradient behavior.
    window.addEventListener('resize', debounce(() => {
      this.headerWidth = this.getHeaderWidth();
      this.calculateOverflow();
      this.updateUnderline();
    }, 500));
    // Check for header overflow on widow resize for gradient behavior.
    this.$refs.cdrTabsHeader.parentElement.addEventListener('scroll', debounce(() => {
      this.calculateOverflow();
      this.updateUnderline();
    }, 50));
  },
  methods: {
    getNextTab(startIndex) {
      for (let i = startIndex; i < this.tabs.length; i += 1) {
        if (!this.tabs[i].disabled) {
          return i;
        }
      }

      if (startIndex !== 0) {
        for (let k = 0; k < startIndex; k += 1) {
          if (!this.tabs[k].disabled) {
            return k;
          }
        }
      }

      return -1;
    },
    getPreviousTab(startIndex) {
      for (let i = startIndex; i > -1; i -= 1) {
        if (!this.tabs[i].disabled) {
          return i;
        }
      }

      if (startIndex !== this.tabs.length - 1) {
        for (let k = this.tabs.length - 1; k > startIndex; k -= 1) {
          if (!this.tabs[k].disabled) {
            return k;
          }
        }
      }

      return -1;
    },
    handleClick: debounce(function handleClickCallback(tabClicked) {
      const newIndex = this.tabs.findIndex((tab) => tabClicked.name === tab.name);
      this.changeTab(newIndex);
    }, 500, { leading: true, trailing: false }),
    changeTab(newIndex) {
      const oldIndex = this.activeTabIndex;

      this.hideScrollBar();
      if (newIndex > oldIndex) {
        this.tabs[oldIndex].setAnimationDirection('exit-left');
        this.tabs[oldIndex].setActive(false);
        setTimeout(() => {
          this.tabs[newIndex].setActive(true);
          this.tabs[newIndex].setAnimationDirection('enter-right');
        }, 200);
      } else {
        this.tabs[oldIndex].setAnimationDirection('exit-right');
        this.tabs[oldIndex].setActive(false);
        setTimeout(() => {
          this.tabs[newIndex].setActive(true);
          this.tabs[newIndex].setAnimationDirection('enter-left');
        }, 200);
      }
      this.activeTabIndex = newIndex;
      this.updateUnderline();
      this.$refs.cdrTabsHeader.children[this.activeTabIndex].focus();
    },
    rightArrowNav: debounce(function handleRightArrow() {
      const nextTab = this.getNextTab(this.activeTabIndex + 1);
      if (nextTab !== -1) {
        this.changeTab(nextTab);
      }
    }, 300, { leading: true, trailing: false }),
    leftArrowNav: debounce(function handleLeftArrow() {
      const previousTab = this.getPreviousTab(this.activeTabIndex - 1);
      if (previousTab !== -1) {
        this.changeTab(previousTab);
      }
    }, 300, { leading: true, trailing: false }),
    calculateOverflow() {
      let containerWidth = 0;
      if (this.$refs.cdrTabsContainer) {
        containerWidth = this.$refs.cdrTabsContainer.offsetWidth;
      }
      this.headerOverflow = this.headerWidth > containerWidth;
      if (this.headerOverflow) {
        // Get Scroll Position
        const scrollX = this.$refs.cdrTabsHeader.parentElement.scrollLeft;
        this.overflowLeft = scrollX > 1;
        this.overflowRight = (scrollX + 1) < (this.headerWidth - containerWidth);
      } else {
        this.overflowLeft = false;
        this.overflowRight = false;
      }
    },
    updateUnderline() {
      const elements = Array.from(this.$refs.cdrTabsHeader.children);
      if (elements.length > 0) {
        const activeTab = elements[this.activeTabIndex];
        const activeRect = activeTab.getBoundingClientRect();
        const parentRect = this.$refs.cdrTabsHeader.getBoundingClientRect();
        const offset = activeRect.x - parentRect.x;

        this.underlineOffsetX = offset
          - this.$refs.cdrTabsHeader.parentElement.scrollLeft;
        this.underlineWidth = activeRect.width;

        // shrink/hide the underline if it scrolls outside the container
        if (this.underlineOffsetX + this.underlineWidth >= parentRect.width) {
          this.underlineWidth = Math.max(0, parentRect.width - this.underlineOffsetX);
          this.underlineOffsetX = Math.min(this.underlineOffsetX, parentRect.width);
        } else if (this.underlineOffsetX < 0) {
          this.underlineWidth = Math.max(0, this.underlineWidth + this.underlineOffsetX);
          this.underlineOffsetX = 0;
        }
      }
    },
    handleDownArrowNav() {
      if (!this.animationInProgress) {
        this.$el.lastElementChild.children[this.activeTabIndex].focus();
      }
    },
    setFocusToActiveTabHeader() {
      this.$refs.cdrTabsHeader.children[this.activeTabIndex].focus();
    },
    getHeaderWidth() {
      let headerElements = [];
      if (this.$refs.cdrTabsHeader) {
        headerElements = Array.from(this.$refs.cdrTabsHeader.children);
      }
      let totalWidth = 0;
      headerElements.forEach((element, i) => {
        // account for margin-left on header elements
        if (i > 0) {
          totalWidth += this.size === 'small' ? Number(CdrSpaceHalfX) : Number(CdrSpaceOneX);
        }
        totalWidth += element.offsetWidth || 0;
      });
      return totalWidth;
    },
    hideScrollBar() {
      const containerRef = this.$refs.cdrTabsContainer.style;
      const slotRef = this.$refs.slotWrapper.style;
      window.addEventListener('transitionend', () => {
        containerRef.setProperty('overflow-x', 'unset');
        slotRef.setProperty('overflow-y', 'unset');
      }, { once: true });
      containerRef.setProperty('overflow-x', 'hidden');
      slotRef.setProperty('overflow-y', 'hidden');
    },
  },
  setup(props, ctx) {
    const tabs = ref([]);
    const underlineOffsetX = ref(0);
    const underlineWidth = ref(0);
    const underlineScrollX = ref(0);
    const activeTabIndex = ref(0);
    const headerWidth = ref(0);
    const headerOverflow = ref(false);
    const overflowLeft = ref(false);
    const overflowRight = ref(false);
    const animationInProgress = ref(false);


    const baseClass = style['cdr-tabs'];
    const underlineStyle = computed(() => {
      return {
        transform: `translateX(${this.underlineOffsetX}px)`,
        width: `${this.underlineWidth}px`,
      };
    });
    const gradientLeftStyle = computed(() => {
      const gradient = `linear-gradient(to left, rgba(255, 255, 255, 0), ${this.backgroundColor})`;
      return {
        background: gradient,
      };
    });
    const gradientRightStyle = computed(() => {
      const gradient = `linear-gradient(to right, rgba(255, 255, 255, 0), ${this.backgroundColor})`;
      return {
        background: gradient,
      };
    });



    return {

    };
  },
};
</script>
