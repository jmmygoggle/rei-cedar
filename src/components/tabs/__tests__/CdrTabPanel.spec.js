import { shallowMount, mount } from '../../../../test/vue-jest-style-workaround.js';
import CdrTabPanel from 'componentdir/tabs/CdrTabPanel';
import CdrTabs from 'componentdir/tabs/CdrTabs';
import Vue from 'vue';


describe('CdrTabPanel', () => {
  it('renders tab', () => {
    const wrapper = mount(CdrTabPanel, {
      propsData: {
        name: 'test',
        id: 'tab1',
        ariaLabelledby: 'tab1',
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('is not active by default', () => {
    const wrapper = shallowMount(CdrTabPanel, {
      propsData: {
        name: 'test',
        id: 'tab1',
        ariaLabelledby: 'tab1',
      },
    });
    expect(wrapper.vm.active).toBe(false);
  });

  it('is active when set', async () => {
    const wrapper = shallowMount(CdrTabPanel, {
      propsData: {
        name: 'test',
        id: 'tab1',
        ariaLabelledby: 'tab1',
      },
    });
    wrapper.vm.setActive(true);
    await wrapper.vm.$nextTick();

    expect(wrapper.vm.active).toBe(true);
  });

  it('set animation direction functions correctly', async () => {
    const wrapper = shallowMount(CdrTabPanel, {
      propsData: {
        name: 'test',
        id: 'tab1',
        ariaLabelledby: 'tab1',
      },
    });
    wrapper.vm.setActive(true);
    await wrapper.vm.$nextTick();

    wrapper.vm.setAnimationDirection('flyLeft');
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.animationDirection).toBe('flyLeft');
  });

  it('set animation direction functions correctly', async () => {
    const wrapper = shallowMount(CdrTabPanel, {
      propsData: {
        name: 'test',
        id: 'tab1',
        ariaLabelledby: 'tab1',
      },
    });
    wrapper.vm.setActive(true);
    await wrapper.vm.$nextTick();

    wrapper.vm.setOffsetX(1234);
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.offsetX).toBe(1234);
  });

  it('updates state after animationend', async () => {
    const wrapper = shallowMount(CdrTabPanel, {
      propsData: {
        name: 'test',
        id: 'tab1',
        ariaLabelledby: 'tab1',
      },
    });

    wrapper.setData({ active: true, hidden: false, animationDirection: 'exit-left' });
    await wrapper.vm.$nextTick();
    wrapper.trigger('animationend', {
      animationName: 'exit-left'
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.hidden).toBe(true);
    expect(wrapper.vm.animationDirection).toBe(null);
  });

  it('handleUpArrowNav', async (done) => {
    const elem = document.createElement('div')
    if (document.body) {
      document.body.appendChild(elem)
    }
    const wrapper = mount(CdrTabs, {
      stubs: {
        'cdr-tab-panel': CdrTabPanel,
      },
      slots: {
        default: ['<cdr-tab-panel name="tab1" id="tab1" aria-labelledby="tab1" />', '<cdr-tab-panel name="tab2" id="tab2" aria-labelledby="tab2" />']
      },
      attachTo: elem,
    });

    const spyUpdateUnderline = spyOn(wrapper.vm, 'updateUnderline');
    const spySetFocusToActiveTabHeader = spyOn(wrapper.vm, 'setFocusToActiveTabHeader');
    await wrapper.vm.$nextTick();

    wrapper.findComponent(CdrTabPanel).trigger('keydown.up');
    expect(spySetFocusToActiveTabHeader).toHaveBeenCalled();
    setTimeout(() => {
      expect(spyUpdateUnderline).toHaveBeenCalled();
      wrapper.destroy();
      done();
    }, 550)
  })
});
