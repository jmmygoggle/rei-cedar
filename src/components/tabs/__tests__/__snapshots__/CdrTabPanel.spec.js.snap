// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdrTabPanel renders tab 1`] = `
<div
  aria-hidden="true"
  aria-labelledby="tab1"
  class="cdr-tab-panel"
  hidden=""
  id="tab1"
  role="tabpanel"
  tabindex="0"
>
  
  
</div>
`;
